# **Detecção do Câncer de Mama**

**Objetivo**
* Criar um algoritmo de machine learning capaz de identificar se um paciente está com câncer de mama.

### **1. Introdução**

<p align="justify">
A detecção precoce do câncer de mama é fundamental para aumentar as chances de tratamento bem-sucedido e melhorar significativamente a qualidade de vida das pacientes. Neste projeto, empregaremos técnicas de Machine Learning, utilizando algoritmos específicos para classificação, a fim de analisar e interpretar as características intricadas presentes nos dados, tais como texturas, dimensões e medidas derivadas de imagens digitalizadas de biópsias de tecido mamário.

<p align="justify">
Ao se debruçar sobre o desafio da classificação binária – se a paciente está ou não com câncer de mama –, nosso objetivo é oferecer uma ferramenta clínica que possa servir como um suporte adicional aos profissionais de saúde no processo de tomada de decisão. Este trabalho não apenas visa a eficácia na identificação de casos positivos, mas também a minimização de falsos positivos, uma vez que a precisão é crucial em contextos médicos.

In [ ]:
# Carregando bibliotecas

import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

In [ ]:
# Carregando base de dados

dados = pd.read_csv('/content/drive/MyDrive/Ciência de Dados com Python/Projetos/Dados/Breast Cancer Wisconsin (Diagnostic) Data Set.csv',
                    sep = ',')

In [ ]:
# visualizando base de dados

dados.head()

Id number Diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0     842302         M        17.99         10.38          122.80     1001.0   
1     842517         M        20.57         17.77          132.90     1326.0   
2   84300903         M        19.69         21.25          130.00     1203.0   
3   84348301         M        11.42         20.38           77.58      386.1   
4   84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave_points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave_points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

### **2. Análise Exploratória de Dados**

In [ ]:
# Verificando a informações dos dados

dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Id number                569 non-null    int64  
 1   Diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave_points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [ ]:
# Verificando a existência de valores faltantes

dados.isnull().sum()

Id number                  0
Diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave_points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave_points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave_points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64

In [ ]:
# Verificando valores únicos

dados['Diagnosis'].unique()

array(['M', 'B'], dtype=object)

### **3.Pré-Processamento de Dados**

In [ ]:
# Alterando valores categóricos para numéricos

dados["Diagnosis"] = dados["Diagnosis"].map({"M": 0, "B": 1})

In [ ]:
# Verificando resultado da transformação e contagem de elementos

dados['Diagnosis'].value_counts()

1    357
0    212
Name: Diagnosis, dtype: int64

In [ ]:
# Separando dados em previsores e classes

previsores = dados.drop('Diagnosis', axis = 1)
classes = dados['Diagnosis']

In [ ]:
# Dividindo os dados em treino e teste

X_treino, X_teste, y_treino, y_teste = train_test_split(previsores,
                                                        classes,
                                                        test_size = 0.3,
                                                        random_state = 0)

In [ ]:
# Distribuição dos dados

print(X_treino.shape)
print(y_treino.shape)
print(X_teste.shape)
print(y_teste.shape)

(398, 31)
(398,)
(171, 31)
(171,)


### **4. Construindo Modelo**

**OBS:** utilizaremos o PyCaret para escolha do modelo mais adequado para o conjunto de dados;

In [ ]:
# Instalando biblioteca

# !pip install pycaret

In [ ]:
# Carregando biblioteca

from pycaret.classification import *

In [ ]:
# Escolhendo o algoritmo mais adequado para o estudo

clf = setup(dados, target = "Diagnosis", session_id = 0)
compare_models()

Processing:   0%|          | 0/65 [00:00<?, ?it/s]

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='sqrt',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_samples_leaf=1,
                     min_samples_split=2, min_weight_fraction_leaf=0.0,
                     n_estimators=100, n_jobs=-1, oob_score=False,
                     random_state=0, verbose=0, warm_start=False)

**OBS 2:** utilizaremos o Extra Trees Classifier para construção do modelo.

In [ ]:
# Carregando biblioteca

from sklearn.ensemble import ExtraTreesClassifier

In [ ]:
# Instânciando o modelo

modelo = ExtraTreesClassifier(n_estimators = 100, random_state = 0)

In [ ]:
# Treinando o modelo

modelo.fit(X_treino, y_treino)

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='sqrt',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_samples_leaf=1,
                     min_samples_split=2, min_weight_fraction_leaf=0.0,
                     n_estimators=100, n_jobs=None, oob_score=False,
                     random_state=0, verbose=0, warm_start=False)

In [ ]:
# Aplicando aos dados de teste

y_predito = modelo.predict(X_teste)

### **5. Avaliando Modelo**

In [ ]:
# Métricas do modelo

print(classification_report(y_teste, y_predito))

              precision    recall  f1-score   support

           0       0.95      0.92      0.94        63
           1       0.95      0.97      0.96       108

    accuracy                           0.95       171
   macro avg       0.95      0.95      0.95       171
weighted avg       0.95      0.95      0.95       171



### **6. Considerações Finais**

<p align="justify">
Os resultados apresentados no classification report refletem um desempenho notável do modelo no contexto da identificação de pacientes com câncer de mama. As métricas de precision, recall e f1-score fornecem uma visão abrangente do desempenho do modelo em ambas as classes.

<p align="justify">
Observamos uma precisão notável para ambas as classes, atingindo 95% para casos negativos (sem câncer) e 95% para casos positivos (com câncer). Isso indica a capacidade do modelo de minimizar falsos positivos e falsos negativos, sendo crucial em contextos clínicos.

<p align="justify">
O recall alcançou 92% para casos negativos e 97% para casos positivos, demonstrando uma habilidade equilibrada do modelo em identificar corretamente pacientes tanto com quanto sem câncer de mama. Este equilíbrio é essencial, garantindo que o modelo não negligencie a detecção de casos positivos, enquanto mantém a precisão em um nível elevado.

<p align="justify">
A acurácia global do modelo atingiu 95%, indicando uma eficácia consistente na classificação de ambas as classes. Essa métrica é especialmente relevante considerando o desequilíbrio natural entre casos positivos e negativos.

<p align="justify">
A média ponderada das métricas (weighted avg) confirma a solidez geral do modelo, com um f1-score de 95%. Esta métrica considera tanto a precisão quanto o recall, proporcionando uma visão equilibrada do desempenho do modelo em ambas as classes.

<p align="justify">
Em conclusão, os resultados indicam que o modelo é eficaz na identificação de casos de câncer de mama, oferecendo uma ferramenta promissora para auxiliar os profissionais de saúde no diagnóstico precoce. Contudo, é sempre crucial validar e aprimorar continuamente o modelo com base em novos dados e avanços tecnológicos, garantindo sua relevância e confiabilidade no cenário clínico em constante evolução.